File: shuffle-array.ts
/*
Given an integer array nums, design an algorithm to randomly shuffle the array. All permutations of the array should be equally likely as a result of the shuffling.

Implement the Solution class:

Solution(int[] nums) Initializes the object with the integer array nums.
int[] reset() Resets the array to its original configuration and returns it.
int[] shuffle() Returns a random shuffling of the array.
*/

class Solution {
    
    private initialConfig: number[];
    private nums: number[];

    constructor(nums: number[]) {
        // console.log(nums)
        this.nums = nums;
        this.initialConfig = [...nums];
    }

    reset(): number[] {
        //this.nums = this.initialConfig;
        return this.initialConfig;
    }

    shuffle(): number[] {
        for(let i = 0; i < this.nums.length; i++){
            const swapIndex = this.switchWith();
            const temp = this.nums[i];
            this.nums[i] = this.nums[swapIndex];
            this.nums[swapIndex] = temp;
        }
        return this.nums
    }

    private switchWith(): number {
        return Math.floor(Math.random() * this.nums.length);
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * var obj = new Solution(nums)
 * var param_1 = obj.reset()
 * var param_2 = obj.shuffle()
 */

File: shuffle-array.cpp
#include <vector>
#include <cstdlib>

class Solution {
public:
    Solution(std::vector<int>& _nums): og(_nums), nums(_nums) {}
    
    std::vector<int> reset() {
      nums = og;
      return nums;
    }
    
    std::vector<int> shuffle() {
      for (int i = 0; i < nums.size(); i++) {
        int randIdx = rand() % nums.size();
        int temp = nums.at(randIdx);
        nums[randIdx] = nums.at(i);
        nums[i] = temp;
      }
    
      return nums;
    }
private:
  const std::vector<int> og;
  std::vector<int> nums;
};
